### ISSUE TITLE
ERROR: resolved path doesnt match with open path

### ISSUE THREAD

**Issue by asura117** [apptainer/singularity#3249]
## Version of Singularity:

3.0.2-1.el7

Hi I am running a nextflow script with singularity but I get the following error message:

Command exit status:
  255

Command output:
  (empty)

Command error:
  FATAL:   resolved path /containers/sarek-latest.simg doesn't match with opened path /containers/sarek-latest.simg (deleted)
  ERROR  : Child exit with status 255

Work dir:
/work/11/9bf34e8a12305c98e2ec87bfdc1e2e

Tip: view the complete command output by changing to the process work dir and entering the command `cat .command.out`
WARN: Killing pending tasks (50)
[b6/480ed9] NOTE: Process `RunFastQC (79887-1)` terminated with an error exit status (255) -- Error is ignored
[c7/c51326] NOTE: Process `RunFastQC (79880-1)` terminated with an error exit status (255) -- Error is ignored
[29/7de678] NOTE: Process `RunFastQC (79890-1)` terminated with an error exit status (255) -- Error is ignored
[11/0cf4a8] NOTE: Process `RunFastQC (79893-1)` terminated with an error exit status (255) -- Error is ignored
[b9/694c31] NOTE: Process `RunFastQC (79894-1)` terminated with an error exit status (255) -- Error is ignored
[94/53d0c5] NOTE: Process `RunFastQC (79882-1)` terminated with an error exit status (255) -- Error is ignored
[ef/d77902] NOTE: Process `RunFastQC (79904-1)` terminated with an error exit status (255) -- Error is ignored
[46/548ef9] NOTE: Process `RunFastQC (79898-1)` terminated with an error exit status (255) -- Error is ignored
[ae/3631f4] NOTE: Process `RunFastQC (79885-1)` terminated with an error exit status (255) -- Error is ignored
[b3/9bc6cc] NOTE: Process `RunFastQC (79895-1)` terminated with an error exit status (255) -- Error is ignored
[1d/d7b428] NOTE: Process `RunFastQC (79919-1)` terminated with an error exit status (255) -- Error is ignored

Is anyone using singularity let me know if this is an error due to Singulairty? If so can anyone suggest a fix.

**Comment by cclerget** [comment]
Hi @asura117 , the error `FATAL: resolved path /containers/sarek-latest.simg doesn't match with opened path /containers/sarek-latest.simg (deleted)` say the image (`/containers/sarek-latest.simg`) was deleted (`/containers/sarek-latest.simg (deleted)`) during singularity execution, which is really weird. Can you please paste a reproducer for this issue ?

**Comment by cclerget** [comment]
Closing, feel free to re-open if you still have this issue @asura117

**Comment by jthiltges** [comment]
One of our users ran across this issue with a SLURM job array. I've been able to reproduce the error with singularity 3.5.2-1.1.osg34.el6 on EL6 (2.6.32-754.27.1.el6.x86_64) by starting several simultaneous singularity instances from the same image.

```
singularity pull docker://godlovedc/lolcow
for i in `seq 1 10` ; do
    singularity run lolcow_latest.sif > /dev/null &
done
wait
```
I was not able to reproduce with singularity 3.5.2-1.1.osg34.el7 on EL7. I suspect making a copy of the sif before starting each job will avoid the problem on EL6.

**Comment by cclerget** [comment]
@jthiltges What is the error message returned, curious if the image is deleted like in the original post ?

**Comment by jthiltges** [comment]
We're indeed seeing the same pattern:

```
FATAL   [U=2111,P=132335]  StageOne()                    resolved path /lustre/work/swanson/jthiltge/fd25442/lolcow_latest.sif doesn't match with opened path /lustre/work/swanson/jthiltge/fd25442/lolcow_latest.sif (deleted)
```

I appreciate your attention. Given that RHEL6 is only about 5 months away from being EOL, this is probably more of a curiosity than something worth spending time fixing.

<details>
<summary>Debug output of a failed run</summary>

```
DEBUG   [U=2111,P=132184]  persistentPreRunE()           Singularity version: 3.5.2-1.1.osg34.el6
DEBUG   [U=2111,P=132184]  handleConfDir()               /home/swanson/jthiltge/.singularity already exists. Not creating.
DEBUG   [U=2111,P=132184]  getCacheBasedir()             environment variable SINGULARITY_CACHEDIR not set, using default image cache
DEBUG   [U=2111,P=132184]  execStarter()                 Checking for encrypted system partition
DEBUG   [U=2111,P=132184]  Init()                        Image format detection
DEBUG   [U=2111,P=132184]  Init()                        Check for sandbox image format
DEBUG   [U=2111,P=132184]  Init()                        sandbox format initializer returned: not a directory image
DEBUG   [U=2111,P=132184]  Init()                        Check for sif image format
DEBUG   [U=2111,P=132184]  Init()                        sif image format detected
VERBOSE [U=2111,P=132184]  SetContainerEnv()             HOME=/home/swanson/jthiltge
DEBUG   [U=2111,P=132184]  init()                        Use starter binary /usr/libexec/singularity/bin/starter-suid
VERBOSE [U=0,P=132184]     print()                       Set messagelevel to: 5
VERBOSE [U=0,P=132184]     init()                        Starter initialization
DEBUG   [U=0,P=132184]     load_overlay_module()         Trying to load overlay kernel module
DEBUG   [U=0,P=132184]     load_overlay_module()         Overlay seems not supported by the kernel
DEBUG   [U=0,P=132184]     get_pipe_exec_fd()            PIPE_EXEC_FD value: 8
VERBOSE [U=0,P=132184]     is_suid()                     Check if we are running as setuid
VERBOSE [U=0,P=132184]     priv_drop()                   Drop root privileges
DEBUG   [U=2111,P=132184]  init()                        Read engine configuration
DEBUG   [U=2111,P=132184]  init()                        Wait completion of stage1
VERBOSE [U=2111,P=132335]  priv_drop()                   Drop root privileges permanently
DEBUG   [U=2111,P=132335]  set_parent_death_signal()     Set parent death signal to 9
VERBOSE [U=2111,P=132335]  init()                        Spawn stage 1
DEBUG   [U=2111,P=132335]  startup()                     singularity runtime engine selected
VERBOSE [U=2111,P=132335]  startup()                     Execute stage 1
DEBUG   [U=2111,P=132335]  StageOne()                    Entering stage 1
DEBUG   [U=2111,P=132335]  prepareAutofs()               Found "/misc" as autofs mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Found "/net" as autofs mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Found "/shared" as autofs mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Found "/shared/unmc1" as autofs mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Found "/cvmfs" as autofs mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Could not keep file descriptor for bind path /etc/localtime: no mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Could not keep file descriptor for bind path /etc/hosts: no mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Could not keep file descriptor for bind path /work: no mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Could not keep file descriptor for bind path /common: no mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Could not keep file descriptor for bind path /lustre: no mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Could not keep file descriptor for home directory /home/swanson/jthiltge: no mount point
DEBUG   [U=2111,P=132335]  prepareAutofs()               Could not keep file descriptor for current working directory /work/swanson/jthiltge/fd25442: no mount point
DEBUG   [U=2111,P=132335]  Init()                        Image format detection
DEBUG   [U=2111,P=132335]  Init()                        Check for sandbox image format
DEBUG   [U=2111,P=132335]  Init()                        sandbox format initializer returned: not a directory image
DEBUG   [U=2111,P=132335]  Init()                        Check for sif image format
DEBUG   [U=2111,P=132335]  Init()                        sif image format detected
FATAL   [U=2111,P=132335]  StageOne()                    resolved path /lustre/work/swanson/jthiltge/fd25442/lolcow_latest.sif doesn't match with opened path /lustre/work/swanson/jthiltge/fd25442/lolcow_latest.sif (deleted)
VERBOSE [U=2111,P=132184]  wait_child()                  stage 1 exited with status 255
```
</details>


<details>
<summary>Partial debug output of a successful run, showing where it diverges</summary>

```
...
DEBUG   [U=2111,P=132354]  Init()                        Image format detection
DEBUG   [U=2111,P=132354]  Init()                        Check for sandbox image format
DEBUG   [U=2111,P=132354]  Init()                        sandbox format initializer returned: not a directory image
DEBUG   [U=2111,P=132354]  Init()                        Check for sif image format
DEBUG   [U=2111,P=132354]  Init()                        sif image format detected
DEBUG   [U=2111,P=132354]  setSessionLayer()             Attempting to use underlay (enable underlay = yes)
VERBOSE [U=2111,P=132185]  wait_child()                  stage 1 exited with status 0
...
```

**Comment by pontus** [comment]
I recently discussed this in a slack as well. There (on a version reporting as 3.0.2-1.el7), they can consistently trigger it. I asked one of them observing it to run a reasonably tight (1 second sleeps)  loop with `stat -c %i FILE` to get the inode as well as observe any issues in accessing the name.

In that situation, the inode did not change (so it was the same file all along) and they saw no issues stating the name (so I believe it was there during the entire run).

This is also with sarek (a nextflow pipeline).

From what I understood from that discussion, it also seems to be related to the amount of singularity processes running simultaneously using that container file (unclear to me whatever that was locally only or in the complete cluster).
